### ISSUE TITLE
TypeError: remoteJQuery is not a function

### ISSUE THREAD

**Issue by James-E-Adams** [cypress-io/cypress#1502]
### Current behavior:
`cy.get(selector)` works fine.
`cy.get(selector).then((elem) => {})` with any callback results in `TypeError: remoteJQuery`, with no extra information
### Desired behavior:

cy.get(selector).then((elem) => {}) should give me access to the element.

#### Test code:
```js
  it('example', () => {
    cy.get('#cy-div')
    .then(elem => {
      elem;
     })
  })
```

I've tried cy v2.0.4 and v2.1

on Mac OS 10.13.3 using chrome.

<img width="406" alt="screen shot 2018-03-27 at 3 46 33 pm" src="https://user-images.githubusercontent.com/9067274/37947310-18a4182a-31d6-11e8-9483-e4fd08ca3b31.png">

**Comment by jennifer-shehane** [comment]
I am unable to reproduce this issue with what you have provided. My working example:

```js
  it('example', () => {
    cy.visit('https://example.cypress.io/commands/actions')
    cy.get('.action-email')
      .then(elem => {
        elem;
      })
  })
```

<img width="663" alt="screen shot 2018-03-27 at 9 15 36 am" src="https://user-images.githubusercontent.com/1271364/37969752-8d2c8e20-319f-11e8-85a1-52422c6732e8.png">

It looks as if you are targeting a `canvas`? Could you provide more detail about the `#cy-div-draw-on` element you're targeting and the behavior you're attempting to test? Or provide a reproducible repository with the issue?

**Comment by James-E-Adams** [comment]
I think I have some weird errors specific to my computer. I'm having no issues running the same tests on other computers. I may have to open a new issue and close this, but the most troubling problem of late is I cannot run more than one test, as it fails like this on the second test in a suite:

<img width="412" alt="screen shot 2018-04-11 at 12 47 56 pm" src="https://user-images.githubusercontent.com/9067274/38593827-a6345b76-3d86-11e8-86ef-2b647c5ab247.png">

My beforeEach looks like: 

```js
  beforeEach(() => {
    cy.authenticatedVisit(url)
  })
```

and the above command is defined in `support/commands` as
```js
Cypress.Commands.add('authenticatedVisit', url => {
  cy.visit(url, {
    onBeforeLoad() {
      window.localStorage.setItem('authRefreshToken', authRefreshToken) 
      window.localStorage.setItem('authAccessToken', null) //it needs this field to have an entry
    },
  })
})
```
`authRefreshToken` is just a string. Has anyone else experienced something similar?

**Comment by vinhlh** [comment]
I have the same issue when baseUrl is from local env https://localhost:3000.

**Comment by James-E-Adams** [comment]
This is resolved with one of the recent updates to Cypress. I can't say which version/why it's fixed, but glad that it is the case. 🎉

**Comment by krasnoperov** [comment]
I have this issue when a string constant with '$' is used somewhere in my code:
`b = v('set'), w = v('append') // pass`
`b = v('$set'), w = v('$append') // fall`

**Comment by marioparris-qless** [comment]
I'm also experiencing this issue (v3.4.0 and v3.4.1). The app -- a very old legacy app -- under test isn't using JQuery but it does define an object `window.$ = { ... }`. Because `window.$` is not a function, I get this error, `remoteJQuery is not a function`, reported. 

It seems to stem from [here](https://github.com/cypress-io/cypress/blob/da497e44d5bdc2f3dd7dbb5834bf0f9a5216aeaa/packages/driver/src/cy/jquery.js#L12)

[Reproducible repo](https://github.com/marioparris-qless/cypress-remoteJQuery)

I'm not sure what I should I do here. Hopefully there's a configuration I can change?

**Comment by XPuigA** [comment]
@marioparris-qless did you manage to solve it? I'm having the same issue.

The webpage does not use jQuery, but it does seem to define window.$ as an empty object. Although I've looked in every dependency JS file from the page and have not found it being set anywhere.

Here is the debugging I've done, in case it helps somebody to see the issue:
![preCall](https://user-images.githubusercontent.com/9863186/66023570-de56ea00-e4f1-11e9-9d03-ab40c01e4843.PNG)
![jquerycall](https://user-images.githubusercontent.com/9863186/66023572-e31b9e00-e4f1-11e9-90a6-c9da2e416194.PNG)
![condition](https://user-images.githubusercontent.com/9863186/66023579-e4e56180-e4f1-11e9-9bcb-244b5d89834b.PNG)
![remoteJqueryNotSameAsGlobal](https://user-images.githubusercontent.com/9863186/66023583-e7e05200-e4f1-11e9-8432-91a009c42f63.PNG)
![callFunctionOnObject](https://user-images.githubusercontent.com/9863186/66023584-ea42ac00-e4f1-11e9-9712-05d0c0d3c645.PNG)
![error](https://user-images.githubusercontent.com/9863186/66023585-eca50600-e4f1-11e9-91ee-be84978ba2d8.PNG)

**Comment by marioparris-qless** [comment]
@XPuigA No, I haven't been able to. You've gotten way further than I have!

**Comment by XPuigA** [comment]
@marioparris-qless I think I fixed it with the following code: 
`cy.on('window:before:load', (win) => {
    if (win.jQuery === undefined && (win.$ === undefined || win.$ === {})) {
        win.$ = Cypress.cy.$$;
    }
});
`
For anyone else having the same problem, what the code does is, before loading every page in the spec, check if the Window object has the jQuery or $ properties, and if not, assign the jQuery that comes with Cypress to Window.$.
As script tags are loaded after this, if the page uses a jQuery library, that one will override the one we set, so the page will still work and we can use jQuery from Cypress.

**Comment by rubenreyes2000** [comment]
I am experiencing the same issue with a site that uses jQuery and jQuery.noConflict() https://api.jquery.com/jQuery.noConflict/

**Comment by marioparris-qless** [comment]
> @marioparris-qless I think I fixed it with the following code:
> 
> ```
> cy.on('window:before:load', (win) => {
>     if (win.jQuery === undefined && (win.$ === undefined || win.$ === {})) {
>         win.$ = Cypress.cy.$$;
>     }
> });
> ```
> 
> For anyone else having the same problem, what the code does is, before loading every page in the spec, check if the Window object has the jQuery or $ properties, and if not, assign the jQuery that comes with Cypress to Window.$.
> As script tags are loaded after this, if the page uses a jQuery library, that one will override the one we set, so the page will still work and we can use jQuery from Cypress.

Nice!

**Comment by kashyaprahul94** [comment]
We were in the similar situation, where our website made use of window.$ as something than jQuery. 

To give you a surprise, this is what we have:
```js
window.$ = document.querySelector.bind(document)
```

It really made us worried initially that we would have to change all the occurrences of $(...) anything to something else in order to make the Cypress work, but luckily, after gloating inside the cypress source code, we found something that could work for us & guess what, it did.

```js
Cypress.on("window:before:load", () => {
  /**
   * Thankfully, Cypress searches for "jQuery" property on the state
   *  variable (cy), if that's present, it takes the precedence
   *  over window.$
   *  https://github.com/cypress-io/cypress/blob/7.0-release/packages/driver/src/cy/jquery.js#L12
   */
  Cypress.cy.state("jQuery", Cypress.$);
});
```

We hope that this could help others too :)

**Comment by jennifer-shehane** [comment]
I can recreate this issue with the code below:

`index.html`

```html
<html>
<body>
  <h1>Hello world</h1>
  <script>
    window.$ = 'foo'
  </script>
</body>
</html>
```

`spec.js`

```js
it('remote jQuery is not a function', () => {
  cy.visit('index.html')
  cy.get('h1').then(() => {})
})
```

<img width="760" alt="Screen Shot 2021-05-05 at 9 17 37 AM" src="https://user-images.githubusercontent.com/1271364/117155849-c22cc600-ad82-11eb-9709-6a442e5c1699.png">

**Comment by awha86** [comment]
> We were in the similar situation, where our website made use of window.$ as something than jQuery.
> 
> To give you a surprise, this is what we have:
> 
> ```js
> window.$ = document.querySelector.bind(document)
> ```
> 
> It really made us worried initially that we would have to change all the occurrences of `$(...)` to something else in order to make the Cypress work, but luckily, after gloating inside the cypress source code, we found something that could work for us & guess what, it did.
> 
> ```js
> Cypress.on("window:before:load", () => {
>   /**
>    * Thankfully, Cypress searches for "jQuery" property on the state
>    *  variable (cy), if that's present, it takes the precedence
>    *  over window.$
>    *  https://github.com/cypress-io/cypress/blob/7.0-release/packages/driver/src/cy/jquery.js#L12
>    */
>   Cypress.cy.state("jQuery", Cypress.$);
> });
> ```
> 
> We hope that this could help others too :)

With Cypress `12.0.0` the need for this workaround is no longer needed for us. Thank you for the solution @kashyaprahul94. 

I wonder if this issue is no longer there, since Cypress `12.0.0`, for anyone else?

**Comment by jennifer-shehane** [comment]
This is still an issue in 13.8.0

![Screenshot 2024-04-23 at 10 14 57 AM](https://github.com/cypress-io/cypress/assets/1271364/6237b3d4-a19d-491f-9efa-c28731086103)
